File: desktop-app/rule_diffs_view_helpers.test.js
// Tests für rule_diffs_view_helpers — pure, node:test-kompatibel.
const test = require("node:test");
const assert = require("node:assert/strict");
const h = require("./rule_diffs_view_helpers");

test("selectPending filtert nur status==='pending' aus project.ruleDiffs", () => {
  const project = {
    ruleDiffs: [
      { id: "a", status: "pending", action: "deactivate", ruleId: "r1", text: "x" },
      { id: "b", status: "approved", action: "activate", ruleId: "r2", text: "y" },
      { id: "c", status: "rejected", action: "deactivate", ruleId: "r3", text: "z" },
    ],
  };
  const out = h.selectPending(project);
  assert.equal(out.length, 1);
  assert.equal(out[0].id, "a");
});

test("selectPending verträgt fehlendes/leeres ruleDiffs", () => {
  assert.deepEqual(h.selectPending(null), []);
  assert.deepEqual(h.selectPending({}), []);
  assert.deepEqual(h.selectPending({ ruleDiffs: [] }), []);
});

test("formatDiffLabel beschreibt action menschlich auf deutsch", () => {
  assert.equal(
    h.formatDiffLabel({ action: "activate", text: "tdd" }),
    "aktivieren: tdd",
  );
  assert.equal(
    h.formatDiffLabel({ action: "deactivate", text: "kein log" }),
    "deaktivieren: kein log",
  );
  assert.equal(
    h.formatDiffLabel({ action: "weird", text: "x" }),
    "änderung: x",
  );
});

test("countPending zählt nur pending", () => {
  const project = {
    ruleDiffs: [
      { status: "pending" }, { status: "pending" }, { status: "approved" },
    ],
  };
  assert.equal(h.countPending(project), 2);
  assert.equal(h.countPending({}), 0);
});

File: mobile-app/test/features/rule_diffs/rule_diffs_view_helpers_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:projectgamma_mobile/features/rule_diffs/rule_diffs_view_helpers.dart';

void main() {
  group('selectPending', () {
    test('leer wenn project null', () {
      expect(selectPending(null), isEmpty);
    });
    test('leer wenn ruleDiffs fehlt', () {
      expect(selectPending(<String, dynamic>{}), isEmpty);
    });
    test('filtert nicht-pending raus', () {
      final p = {
        'ruleDiffs': [
          {'id': '1', 'status': 'pending', 'text': 'a'},
          {'id': '2', 'status': 'approved', 'text': 'b'},
          {'id': '3', 'status': 'rejected', 'text': 'c'},
          {'id': '4', 'status': 'pending', 'text': 'd'},
        ],
      };
      final out = selectPending(p);
      expect(out, hasLength(2));
      expect(out.map((d) => d['id']), containsAll(['1', '4']));
    });
  });

  group('countPending', () {
    test('liefert 0 bei null/empty', () {
      expect(countPending(null), 0);
      expect(countPending(<String, dynamic>{}), 0);
    });
    test('zählt nur pending', () {
      final p = {
        'ruleDiffs': [
          {'status': 'pending'}, {'status': 'pending'}, {'status': 'approved'},
        ],
      };
      expect(countPending(p), 2);
    });
  });

  group('formatDiffLabel', () {
    test('null → leer', () {
      expect(formatDiffLabel(null), '');
    });
    test('activate-action', () {
      expect(formatDiffLabel({'action': 'activate', 'text': 'foo'}),
          'aktivieren: foo');
    });
    test('deactivate-action', () {
      expect(formatDiffLabel({'action': 'deactivate', 'text': 'foo'}),
          'deaktivieren: foo');
    });
    test('unbekannte action → fallback', () {
      expect(formatDiffLabel({'action': 'wat', 'text': 'foo'}),
          'änderung: foo');
    });
  });
}
